File: app/src/main/java/beans/prj/newsapp/ui/base/ViewModelProviderFactory.java
package beans.prj.newsapp.ui.base;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import beans.prj.newsapp.data.AppRepository;

import java.lang.reflect.InvocationTargetException;

public class ViewModelProviderFactory implements ViewModelProvider.Factory {

    AppRepository repository;

    public ViewModelProviderFactory(AppRepository repository) {
        this.repository = repository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (ViewModel.class.isAssignableFrom(modelClass)) {
            try {
                return modelClass.getConstructor(AppRepository.class).newInstance(repository);
            } catch (NoSuchMethodException e) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, e);
            } catch (IllegalAccessException e) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, e);
            } catch (InstantiationException e) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, e);
            } catch (InvocationTargetException e) {
                throw new RuntimeException("Cannot create an instance of " + modelClass, e);
            }
        }
        return null;
    }
}

File: app/src/main/java/beans/prj/newsapp/ui/base/ViewModelProviderFactory.kt
package beans.prj.newsapp.ui.base

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import beans.prj.newsapp.data.AppRepository
import java.lang.reflect.InvocationTargetException

class ViewModelProviderFactory(var repository: AppRepository?) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T = /*if (ViewModel::class.java.isAssignableFrom(modelClass)) */
            modelClass.getConstructor(AppRepository::class.java).newInstance(repository)
    /* else null*/
}